### DISCUSSION TITLE
local-ref assignment - initialised with Unsafe.NullRef

### DISCUSSION THREAD

**Discussion by bernd5** [dotnet/csharplang#4547] [General]
For low level code, working with ref vars is much nicer compared to raw pointers - so I prefer to use ref.
I encountered recently the need to use `Unsafe.NullRef<T>` - which is very nice.

My problem is now, that the "safe to escape rules" apply to `Unsafe.NullRef<T>`, too.
Would it be possible to introduce an exception for example by an attribute to allow the following sample code:

### Sample:
```CSharp
ref int i = ref System.Runtime.CompilerServices.Unsafe.NullRef<int>();
int someData = 4711;
i = ref someData; //<--- Cannot ref-assign 'someData' to 'i' because 'someData' has a narrower escape scope than 'i'

System.Console.WriteLine(i);
```

I could write it like:
```CSharp
unsafe
{
    int* i = null;
    int someData = 4711;
    i = &someData;

    System.Console.WriteLine(*i);
}
```
But `ref` is very much nicer!

**Comment by huoyaoyuan** [comment]
Since you are already unsafe here, `Unsafe.AsRef(in variable)` or `Unsafe.Add(ref variable, 0)` may work for you.

**Comment by bernd5** [comment]
What works is the following:

```CSharp
ref T GetNullRef<T>(ref T t)  => ref System.Runtime.CompilerServices.Unsafe.NullRef<T>();

int dummy = 12;
ref int i = ref GetNullRef(ref dummy);
int someData = 4711;
i = ref someData;

System.Console.WriteLine(i);
```

But I don't like such "dummy" variables and the need to introduce a function with an unused parameter...

**Comment by bernd5** [comment]
I found an even better way to achieve the desired behaviour - it uses an "invisible" local dummy:

```CSharp
ref T GetNullRef<T>(in int dummy = default)  => ref System.Runtime.CompilerServices.Unsafe.NullRef<T>();

ref int i = ref GetNullRef<int>();
int someData = 4711;
i = ref someData;

System.Console.WriteLine(i);
```

Should this issue moved to runtime?

I created an issue there: [Link](https://github.com/dotnet/runtime/issues/49808)
